presto-hive/src/main/java/com/facebook/presto/hive/StoragePartitionLoader.java

```diff
@@ -0,0 +1,583 @@
+            boolean splittable,
+            PathFilter pathFilter)
+    {
+        int readBucketCount = bucketSplitInfo.getReadBucketCount();
+        int tableBucketCount = bucketSplitInfo.getTableBucketCount();
+        int partitionBucketCount = bucketConversion.map(HiveSplit.BucketConversion::getPartitionBucketCount).orElse(tableBucketCount);
+        int bucketCount = max(readBucketCount, partitionBucketCount);
+
+        checkState(readBucketCount <= tableBucketCount, "readBucketCount(%s) should be less than or equal to tableBucketCount(%s)", readBucketCount, tableBucketCount);
+
+        // list all files in the partition
+        List<HiveFileInfo> fileInfos = new ArrayList<>(partitionBucketCount);
+        try {
+            Iterators.addAll(fileInfos, directoryLister.list(fileSystem, table, path, namenodeStats, pathFilter, new HiveDirectoryContext(FAIL, isUseListDirectoryCache(session))));
+        }
+        catch (HiveFileIterator.NestedDirectoryNotAllowedException e) {
+            // Fail here to be on the safe side. This seems to be the same as what Hive does
+            throw new PrestoException(
+                    HIVE_INVALID_BUCKET_FILES,
+                    format("Hive table '%s' is corrupt. Found sub-directory in bucket directory for partition: %s",
+                            table.getSchemaTableName(),
+                            partitionName));
+        }
+
+        ListMultimap<Integer, HiveFileInfo> bucketToFileInfo = ArrayListMultimap.create();
+
+        if (!shouldCreateFilesForMissingBuckets(table, session)) {
+            fileInfos.stream()
+                    .forEach(fileInfo -> {
+                        String fileName = fileInfo.getPath().getName();
+                        OptionalInt bucket = getBucketNumber(fileName);
+                        if (bucket.isPresent()) {
+                            bucketToFileInfo.put(bucket.getAsInt(), fileInfo);
+                        }
+                        else {
+                            throw new PrestoException(HIVE_INVALID_BUCKET_FILES, format("invalid hive bucket file name: %s", fileName));
+                        }
+                    });
+        }
+        else {
+            // build mapping of file name to bucket
+            for (HiveFileInfo file : fileInfos) {
+                String fileName = file.getPath().getName();
+                OptionalInt bucket = getBucketNumber(fileName);
+                if (bucket.isPresent()) {
+                    bucketToFileInfo.put(bucket.getAsInt(), file);
+                    continue;
+                }
+
+                // legacy mode requires exactly one file per bucket
+                if (fileInfos.size() != partitionBucketCount) {
+                    throw new PrestoException(
+                            HIVE_INVALID_BUCKET_FILES,
+                            format("Hive table '%s' is corrupt. File '%s' does not match the standard naming pattern, and the number " +
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format int: &#39;size&#39; expression, partitionBucketCount (declared at line 348). This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 396. Consider formatting this data with &#39;%d&#39; instead.